istio.mixer.adapter.model.v1beta/grpc_adapter_service.proto

```diff
@@ -0,0 +1,24 @@
+syntax = "proto3";
+
+// Template agnostic service interface.
+package istio.mixer.adapter.model.v1beta;
+
+
+message SetAdapterConfigReq {
+    map<string, google.protobuf.Value> allHandlerConfigsByName = 1;
+}
+
+message ValidateReq {
+    string handlerName = 1;
+}
+
+message BuildReq {
+    string handlerName = 1;
+}
+
+service AdapterService {
+    rpc  SetAdapterConfig(SetAdapterConfigReq) returns (google.rpc.Status);
```

Comment:
I think this protocol needs to overall learn/use ids for communication. Any place where Mixer passes a pointer to something, this protocol needs to use id strings instead.

I think you start with NewBuilder that returns a build ID.
With this ID, you can call SetXXX and Build.
Build returns a new ID
With this new ID, you call HandleXXX and Close.

This assumes that the target gRPC endpoint is stateful such that the IDs can be shared across nodes. It presumes that the adapter logic will have some sort of garbage collection to clean up unused ids (in case of crash), etc. We also assume that the adapter can dedup, so when presented the same config, will return the same id.

If we use this API to build "near" adapters, whose process is managed by Mixer, then the adapter doesn't need storage, doesn't need garbage collection, doesn't need deduping.

I thus contend that supporting near adapters and having Mixer fully manage the adapter processes is considerably easier to support within the adapters proper.

Reply:
If I understand right, you are proposing a slight variation than the proposed interface here. In your proposal we keep the exact same interface as today's Go adapter interface, except change objects with Ids. May be that one will be more readable, I will update this PR, and we can take a look.

+1 to managing a stateful adapter service with this ID based stuff if going to be non trivial. And as we understand, the other flip side of fully stateless comes at a cost of performance during the critical request time (transferring state info on each call + doing significant work on each request as the state couldn't be cached early on). I think what we can do for adapter service is to provide a shim for just Go language (nice to have in the future), which they can use and that shim will take care of all the garbage collection, deduping etc.
Also note, we will definitely need a shim (for the first release of gRPC adapter model) for "near" adapters and the shim will implement the same interface as the gRPC adapter interface but talk with existing adapters via the existing go interface.

Reply:
That is one reason that Google APIs always use `name` to refer to a resource, and almost all operations apply to a resource or multiple resources. I agree it would be useful pattern here.